OpenRA.Mods.Common/Traits/Buildings/FreeActor.cs

```diff
@@ -28,25 +28,49 @@ public class FreeActorInfo : ITraitInfo
 		[Desc("Which direction the unit should face.")]
 		public readonly int Facing = 0;
 
-		public virtual object Create(ActorInitializer init) { return new FreeActor(init, this); }
+		[Desc("Whether another actor should spawn upon re-enabling the trait.")]
+		public readonly bool AllowRespawn = false;
+
+		public override object Create(ActorInitializer init) { return new FreeActor(init, this); }
 	}
 
-	public class FreeActor
+	public class FreeActor : ConditionalTrait<FreeActorInfo>
 	{
-		public FreeActor(ActorInitializer init, FreeActorInfo info)
+		FreeActorInfo info;
+		bool allowSpawn = true;
+
+		public FreeActor(ActorInitializer init, FreeActorInfo info) : base(info)
 		{
 			if (init.Contains<FreeActorInit>() && !init.Get<FreeActorInit>().ActorValue)
 				return;
 
-			init.Self.World.AddFrameEndTask(w =>
+			this.info = info;
+
+			if (!IsTraitDisabled)
+				GrantFreeActor(init.Self);
+		}
+
+		protected override void TraitEnabled(Actor self)
+		{
+			GrantFreeActor(self);
+		}
+
+		void GrantFreeActor(Actor self)
+		{
+			if (!allowSpawn)
+				return;
+
+			self.World.AddFrameEndTask(w =>
 			{
 				w.CreateActor(info.Actor, new TypeDictionary
 				{
-					new ParentActorInit(init.Self),
-					new LocationInit(init.Self.Location + info.SpawnOffset),
-					new OwnerInit(init.Self.Owner),
+					new ParentActorInit(self),
+					new LocationInit(self.Location + info.SpawnOffset),
+					new OwnerInit(self.Owner),
 					new FacingInit(info.Facing),
 				});
+
+				allowSpawn = info.AllowRespawn;
```

Comment:
This should probably be set outside the AddFrameEndTask, since technically we can still have multiples if it is granted multiple times in the same frame.

Not sure if the implementation for the granting only allows it once a frame but we probably should not depend on implementation here and just move it out.

Reply:
Good spotting, agreed.